### PR TITLE
[wip] cli: Add info command

### PR DESCRIPTION
**Pull request by aequitas**
For now just lists resources and functions, more to be added soon.

```
$ ./mgmt info funcs -t
answer			answer() int
contains
datetime		datetime() int
datetime_print		datetime_print(int) str
defaultenv		defaultenv(str, str) str
env			env() {str: str}
example_errorbool	example_errorbool(bool) str
exchange		exchange(str, str) {str: str}
flipflop		flipflop() bool
getenv			getenv(str) str
hasenv			hasenv(str) bool
hostname		hostname() str
int2str			int2str(int) str
kvlookup		kvlookup(str) {str: str}
len
load			load() struct{x1 float; x5 float; x15 float}
maplookup
pow			pow(float, float) float
printf
random1			random1(int) str
schedule
structlookup
template
vumeter			vumeter(str, int, float) str
```
```
$ ./mgmt info resources
augeas
aws:ec2
exec
file
graph
group
hostname
kv
msg
noop
nspawn
password
pkg
print
svc
test
timer
user
virt
```

### CODE REVIEW

**Comment by aequitas** [comment]
fixes https://github.com/purpleidea/mgmt/issues/312

**Comment by aequitas** [comment]
Now also with a little bit of resource types:

```
augeas    file (string), lens (string), sets ([]resources.AugeasSet)
aws:ec2   state (string), region (string), type (string), imageid (string), watchendpoint (string), watchlistenaddr (string), erroronmalformedpost (bool), userdata (string), publicipv4 (string), privateipv4 (string), instanceid (string)
exec      cmd (string), shell (string), timeout (int), watchcmd (string), watchshell (string), ifcmd (string), ifshell (string), user (string), group (string), output (*string), stdout (*string), stderr (*string)
file      path (string), dirname (string), basename (string), content (*string), source (string), state (string), owner (string), group (string), mode (string), recurse (bool), force (bool)
graph     graph (*pgraph.Graph)
group     state (string), gid (*uint32)
hostname  hostname (string), prettyhostname (string), statichostname (string), transienthostname (string)
kv        key (string), value (*string), skiplessthan (bool), skipcmpstyle (resources.KVResSkipCmpStyle)
msg       body (string), priority (string), fields (map[string]string), journal (bool), syslog (bool)
noop      comment (string)
nspawn    state (string)
password  length (uint16), saved (bool), checkrecovery (bool), password (*string)
pkg       state (string), allowuntrusted (bool), allownonfree (bool), allowunsupported (bool)
print     msg (string)
svc       state (string), startup (string), session (bool)
timer     interval (uint32)
user      state (string), uid (*uint32), gid (*uint32), group (*string), groups ([]string), homedir (*string), allowduplicateuid (bool)
virt      uri (string), state (string), transient (bool), cpus (uint), maxcpus (uint), memory (uint64), osinit (string), boot ([]string), disk ([]resources.diskDevice), cdrom ([]resources.cdRomDevice), network ([]resources.networkDevice), filesystem ([]resources.filesystemDevice), auth (*resources.VirtAuth), hotcpus (bool), restartondiverge (string), restartonrefresh (bool)
```

It's ugly atm, eg: no proper indicator for readonly types.

File: lib/info.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+var myPrintln = fmt.Printf
+
+// info wraps infocmd to produce output to stdout
+func info(c *cli.Context) error {
+	output, err := infocmd(c)
+	if err != nil {
+		return err
+	}
+	w := tabwriter.NewWriter(os.Stdout, 0, 8, 2, ' ', 0)
```
Comment (purpleidea) [inline lib/info.go]: Can we have these vars be `const` (at the top), if only to improve readability please?

File: lib/info.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	"strings"
+	"text/tabwriter"
+
+	"github.com/purpleidea/mgmt/lang/funcs"
+	"github.com/purpleidea/mgmt/lang/interfaces"
+	"github.com/purpleidea/mgmt/resources"
+	"github.com/urfave/cli"
+)
+
+var myPrintln = fmt.Printf
```
Comment (purpleidea) [inline lib/info.go]: what's this?

File: lib/info.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+// infocmd takes the cli context and returns the requested output
+func infocmd(c *cli.Context) (bytes.Buffer, error) {
+	var out bytes.Buffer
+	var names []string
+	descriptions := make(map[string]string)
+
+	switch c.Command.Name {
+	case "resources":
+		for _, name := range resources.RegisteredResourcesNames() {
+			if name == "test" {
```
Comment (purpleidea) [inline lib/info.go]: why skip it? It's real :)

File: lib/info.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+			res, err := resources.Lookup(name)
+			if err != nil {
+				continue
+			}
+
+			s := reflect.ValueOf(res).Elem()
+			typeOfT := s.Type()
+
+			var fields []string
```
Comment (purpleidea) [inline lib/info.go]: use, improve, or add to one of the functions in `resources/util.go` please,

if anything wouldn't we want the info sub commands to be separate functions?

File: lib/info.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+			if strings.HasPrefix(name, "_") {
+				continue
+			}
+			names = append(names, name)
+			descriptions[name] = ""
+			fn, err := funcs.Lookup(name)
+			if err != nil {
+				continue
+			}
+			switch fn.(type) {
```
Comment (purpleidea) [inline lib/info.go]: `_, ok := fn.(interfaces.PolyFunc)`

File: lib/info_test.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	ctx.Command = cli.Command{Name: "functions"}
+
+	out, err := infocmd(ctx)
+	if err != nil {
+		t.Fatal("failed")
+	}
+	assert.Contains(t, out.String(), "load() struct{x1 float; x5 float; x15 float}")
+}
+
+// func TestPolyFunctionsWithTypes(t *testing.T) {
```
Comment (purpleidea) [inline lib/info_test.go]: add or delete?

File: resources/resources.go [hunk 7]
```diff
@@ -673,3 +673,12 @@ func VtoR(v pgraph.Vertex) Res {
 //	}
 //	return nil
 //}
+
+// Lookup returns a pointer to the resource's struct.
```
Comment (purpleidea) [inline resources/resources.go]: Not sure why we need a helper function for this.

File: lib/info.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+	if err != nil {
+		return err
+	}
+	w := tabwriter.NewWriter(os.Stdout, 0, 8, 2, ' ', 0)
+	fmt.Fprint(w, output.String())
+	w.Flush()
+	return nil
+}
+
+// infocmd takes the cli context and returns the requested output
```
Comment (purpleidea) [inline lib/info.go]: golang style says `InfoCmd` or `infoCmd` please, but not sure why this is a separate function altogether

File: lib/info.go (hunk 1 above)
Comment (aequitas) [reply lib/info.go]: sure

File: lib/info.go (hunk 2 above)
Comment (aequitas) [reply lib/info.go]: leftover from the wrong approach to capture test output

File: lib/info_test.go (hunk 6 above)
Comment (aequitas) [reply lib/info_test.go]: scratchpad until i have polyfuncs figured out or for a 2.0 patch of this feature

File: resources/resources.go (hunk 7 above)
Comment (aequitas) [reply resources/resources.go]: Just mirroring funcs Lookup here

File: lib/info.go (hunk 8 above)
Comment (aequitas) [reply lib/info.go]: to allow unit testing the output with jumping through hoops to capture it (eg: rerouting stdout, etc)

**Comment by aequitas** [comment]
@purpleidea 

1: the resource rending is more a hack than a preferred implementation, if you know of a way to get the right output i'm all ears

2: if there is an easy library which can write this to XML in a open format suitable for this kind of data I will add it in, otherwise next patch.

File: resources/resources.go (hunk 7 above)
Comment (aequitas) [reply resources/resources.go]: also registeredResources is private :)

Comment (purpleidea) [reply resources/resources.go]: I didn't realize this was in resources.go :P

**Comment by purpleidea** [comment]
@aequitas Indeed, there is already such an XML lib as seen in yamlgraph2/ being used.

**Comment by aequitas** [comment]
@purpleidea I don't see any xml reference in yamlgraph2? Can you point me to where it is?

**Comment by purpleidea** [comment]
@aequitas Sorry, by brain accidentally some letters. Yaml*:

https://github.com/purpleidea/mgmt/blob/master/yamlgraph2/gconfig.go#L31

**Comment by aequitas** [comment]
@purpleidea so you want xml output or yaml output? :) Why not both? We could maybe use YAML as an intermediate format for XML (and JSON for that matter :D)

**Comment by purpleidea** [comment]
@aequitas Just yaml is sufficient :)

File: lib/info.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+	"strings"
+	"text/tabwriter"
+
+	"github.com/purpleidea/mgmt/lang/funcs"
+	"github.com/purpleidea/mgmt/lang/interfaces"
+	"github.com/purpleidea/mgmt/resources"
+	"github.com/urfave/cli"
+)
+
+const (
```
Comment (purpleidea) [inline lib/info.go]: https://golang.org/doc/effective_go.html#mixed-caps

probably these could be public if you want.

Comment (aequitas) [reply lib/info.go]: i know, i'm following the names set by the tabwriter package, let me know if you prefer proper go names instead

Comment (purpleidea) [reply lib/info.go]: oh gosh. Another inconsistency in their own style guide? Gah. Sorry. Do the right thing. If golangv2 fixes these things, we'll be ahead of the curve.

Comment (aequitas) [reply lib/info.go]: fixed

**Comment by aequitas** [comment]
https://github.com/purpleidea/mgmt/pull/368#issuecomment-372639491

**Pull request closed without merging**